packages/core/contracts/oracle/implementation/OptimisticOracle.sol

```diff
@@ -0,0 +1,368 @@
+        Proposed, // Proposed, but not expired or disputed yet.
+        Expired, // Proposed, not disputed, past liveness.
+        Disputed, // Disputed, but no DVM price returned yet.
+        Resolved, // Disputed and DVM price is available.
+        Settled // Final price has been set in the contract (can get here from Expired or Resolved).
+    }
+
+    struct Request {
+        address proposer;
+        address disputer;
+        IERC20 currency;
+        bool settled;
+        bool refundOnDispute;
+        int256 proposedPrice;
+        int256 resolvedPrice;
+        uint256 expirationTime;
+        uint256 reward;
+        uint256 finalFee;
+        uint256 bond;
+        uint256 customLiveness;
+    }
+
+    mapping(bytes32 => Request) public requests;
+
+    FinderInterface public finder;
+    uint256 public defaultLiveness;
+
+    constructor(
+        uint256 _liveness,
+        address _finderAddress,
+        address _timerAddress
+    ) public Testable(_timerAddress) {
+        finder = FinderInterface(_finderAddress);
+        _validateLiveness(_liveness);
+        defaultLiveness = _liveness;
+    }
+
+    function requestPrice(
+        bytes32 identifier,
+        uint256 timestamp,
+        IERC20 currency,
+        uint256 reward
+    ) external nonReentrant() returns (uint256 totalBond) {
+        require(getState(msg.sender, identifier, timestamp) == State.Invalid, "requestPrice: Invalid");
+        require(_getIdentifierWhitelist().isIdentifierSupported(identifier), "Unsupported identifier");
+        require(_getCollateralWhitelist().isOnWhitelist(address(currency)), "Unsupported currency");
+        uint256 finalFee = _getStore().computeFinalFee(address(currency)).rawValue;
+        requests[_getId(msg.sender, identifier, timestamp)] = Request({
+            proposer: address(0),
+            disputer: address(0),
+            currency: currency,
+            settled: false,
+            refundOnDispute: false,
+            proposedPrice: 0,
+            resolvedPrice: 0,
+            expirationTime: 0,
+            reward: reward,
+            finalFee: finalFee,
+            bond: finalFee,
+            customLiveness: 0
+        });
+
+        if (reward > 0) {
+            currency.safeTransferFrom(msg.sender, address(this), reward);
+        }
+
+        emit RequestPrice(msg.sender, identifier, timestamp);
+
+        return finalFee.mul(2);
+    }
+
+    function setBond(
+        bytes32 identifier,
+        uint256 timestamp,
+        uint256 bond
+    ) external nonReentrant() returns (uint256 totalBond) {
+        require(getState(msg.sender, identifier, timestamp) == State.Requested, "setBond: Requested");
+        Request storage request = _getRequest(msg.sender, identifier, timestamp);
+        request.bond = bond;
+        return bond.add(request.finalFee);
+    }
+
+    function setRefundOnDispute(bytes32 identifier, uint256 timestamp) external nonReentrant() {
```

Comment:
This also seems deprecated, I'm going to stop commenting on this contract now since it seems like there are still local changes to be pushed!